### ISSUE TITLE
Network errors not handled properly

### ISSUE THREAD

**Issue by ryan-summers** [quartiq/minimq#37]
When a network error occurs, the currently-active socket may be inadvertently dropped, as pointed out in https://github.com/quartiq/minimq/pull/24#issuecomment-792432265

**Comment by HarryMakes** [comment]
At the time of writing [my comment on #24](https://github.com/quartiq/minimq/pull/24#issuecomment-792432265), there is such a concern because `socket_is_connected()` didn't consider the case where the socket is `None()`. This would happen when `read()`/`write()` returns an error due to some network error (e.g. disconnection), where the socket reference after being `take()`-n would be left `None()`, meaning the next polling on the client would panic when trying to unwrap the socket by calling `socket_is_connected()`.

Since #35 has been merged, there shouldn't be any further issue of having the socket dropped when `read()`/`write()` raises error. When polling, the client would call the updated `socket_is_connected()`, which guarantees to replace the empty socket reference with an open socket wrapped as `Some()`. Therefore, I believe the active socket can now no longer be dropped from the memory inadvertently.

**Comment by HarryMakes** [comment]
But, I'm still interested to know why #24 removes the blocking mechanism in `read()` but keeps it in `write()`. Would there be any inadvertent wait on writes, and should I create a new issue based on this question?

**Comment by ryan-summers** [comment]
I agree that we shouldn't be blocking on writes either, but I need to think about the implications of not being able to write to the network port.

If we don't block, we have to internally cache a received message for the future, which may be possible. I'll investigate the next steps.

**Comment by ryan-summers** [comment]
Closing this issue in favor of https://github.com/quartiq/minimq/issues/38, as the original bug described in this issue was fixed in #35
